src/thor/trace_route_action.cc

```diff
@@ -455,33 +440,67 @@ thor_worker_t::map_match(Api& request) {
                                         });
         std::get<2>(edge_group) = match_result_itr;
 
-        // TODO:: handle case when both discontinuity and loop on last edge
         // we know the discontinuity happens on this edge so we need to
-        // find the first match result on this edge
-        match_result_itr = std::find_if(std::next(match_result_itr), match_results.cend(),
-                                        [last_edge](const MatchResult& result) {
-                                          return result.edgeid == last_edge->edgeid;
-                                        });
+        // find the last match result on this edge
+        // when the first edge and the last edge has the same edge id, we have to move the edge iter
+        // over the first edge (if there is no discontinuity suggested by the distance along), then
+        // we are able to find the match result on the last edge by std::find_if
```

Comment:
so they have the same edge id but they arent the same edge in the path (ie loop). its hard to understand this from the comment. the thing in parenthesis should say this explicitly as it would very much aide in understanding what is going on. also you ditched std::find_if in favor of a while loop so you should probably update that

Reply:
so if we have a loop we need to push the prev_match_result forward by 1 so that in the discontinuity loop below we can stop at the right place and not get tricked into stopping early?

Reply:
I re-engineered that a little.
There 1 thing we need to know to find the last matched result correctly, which is wether there is a loop on the last edge. 
If there is none, we just need to find the last matched result with the same edgeid here.
If there is loop, we need to find which one is the last matched result by comparing distance along.